tdms/src/interpolation_methods.cpp

```diff
@@ -0,0 +1,394 @@
+}
+
+/**
+ * @brief Determines the appropriate interpolation scheme to use
+ *
+ * @param[in] cells_in_direction The number of Yee cells in the interpolation direction of interest
+ * @param[in] cell_id The current ID (in this dimension) of the Yee cell
+ * @return interp_scheme Indicating the appropriate interpolation scheme to use
+ */
+interp_scheme determineInterpScheme(int cells_in_direction, int cell_id)
+{
+    // interpolation is impossible if the total number of cells in this direction is <4
+    if (cells_in_direction < 4)
+    {
+        throw out_of_range("Error: computational domain has fewer than 4 cells in at least 1 dimension, cubic and band-limited interpolation impossible.\n");
+    }
+    // Yee cell with index <0 doesn't exist (indexing starts from 1), and Yee cell 0 cannot undergo interpolation to the centre
+    else if (cell_id <= 0)
+    {
+        throw out_of_range("Error: Yee cell index <=0 requested (must be >=1).\n");
+    }
+    // Yee cell with index >=cells_in_direction doesn't exist
+    else if (cell_id >= cells_in_direction)
+    {
+        throw out_of_range("Error: requested Yee cell index beyond maximum number of Yee cells.\n");
+    }
+    else if (cells_in_direction < 8)
+    {
+        // we do not have enough cells to use bandlimited interpolation, but we can use cubic
+        // by definition, cell_id = 1 requires us to use interp2, cell_id = cells_in_direction-1 interp3,
+        // and everything else interp2
+        if (cell_id == 1)
+        {
+            return CUBIC_INTERP_FIRST;
+        }
+        else if (cell_id == (cells_in_direction - 1))
+        {
+            return CUBIC_INTERP_LAST;
+        }
+        else
+        {
+            return CUBIC_INTERP_MIDDLE;
+        }
+    }
+    else
+    {
+        // we can apply bandlimited interpolation
+        if ((cell_id >= 4) && (cell_id <= cells_in_direction - 4))
+        {
+            return BAND_LIMITED_3;
+        } // best, and most frequent, case
+        else if (cell_id == 1)
+        {
+            return BAND_LIMITED_0;
+        }
+        else if (cell_id == 2)
+        {
+            return BAND_LIMITED_1;
+        }
+        else if (cell_id == 3)
+        {
+            return BAND_LIMITED_2;
+        }
+        else if (cell_id == cells_in_direction - 3)
+        {
+            return BAND_LIMITED_4;
+        }
+        else if (cell_id == cells_in_direction - 2)
+        {
+            return BAND_LIMITED_5;
+        }
+        else
+        {
+            return BAND_LIMITED_6;
+        } // cell_id = cells_in_direction-1
+    }
+}
+
+/**
+ * @brief Given two interpolation schemes s0 and s1, performs evaluation of the expression "s0 > s1", in terms of optimality of the scheme.
+ *
+ * @param s0,s1 interp_schemes
+ * @return true If s0 is the optimal scheme.
+ * @return false If s1 is the optimal scheme. Also returned in the event of a tie.
+ */
+bool better_scheme(interp_scheme s0, interp_scheme s1) {
```

Comment:
This may be slightly more readable if the interp_scheme was a class, then you can have a method like `scheme0.is_better_than(scheme1)`

Reply:
I debated it; went with an enum because besides this one function, I don't actually need any fancy functionality surrounding `interp_scheme`. Can convert it to a class if you'd prefer.

Reply:
it might be worth it – as they can also then have their own interp_data attributes that are current defined at the top of each `interpolateTimeDomainE{x,y,z}` function (`{cubic,bandlimited}_interpolation` can also take 1 fewer argument)

Reply:
Will make the class and push the changes tomorrow morning :+1:

Reply:
Class now exists: https://github.com/UCL/TDMS/blob/5f470ae405af6f6cff8d520eaf8faf5d8a67466a/tdms/include/interpolation_methods.h#L92

Reply:
having both `interpScheme` and `interp_scheme` with the latter generated with `scheme.get_scheme()` feels weird 😕 
I was thinking to have some const class instances with associated _values_ e.g

```c++
const auto BAND_LIMITED_0 = interpScheme(0);   // where 0 is the value
```
which then have a overloaded call `()` operator to evaluate whatever interpolation they do. Then a function like

```c++
interpScheme best_interpolation_scheme(int cells_in_direction, int cell_id){
...
}
```

apologies if this is foolish and not possible 😅

Reply:
It's probably possible, but it boils down to where we want the "determine the scheme" part to come into it. It has been a while since I got this deep into C++ classes and overloads though :sweat_smile:.

The `interp_scheme` enum is there to make sure we can't request a scheme that doesn't exist, and just to make things a bit more readable than "use scheme 0". I guess we don't need it; the implicit converter/overloaded constructor method will still work if you pass it an `int`, and I can just add an error check if an invalid scheme is requested.

In its current state, creating an instance of the class immediately determines the interpolation scheme it will perform, which can then be run with the `.interpolate(double *v)` method. This means when we interpolate we do the following:
- Decide on a cell to interpolate
- Create an instance of the `interScheme` class - this also determines the best interpolation scheme and sets this class instance to use that scheme
- Gather the interpolation data we need
- Pass the data to `.interpolate()` (or an overloaded call operator).

With constant class instances all we're doing to do is move the "determine which scheme" part:
- Decide on a cell to interpolate
- Determine the appropriate scheme to use, and create a pointer (or similar) to the corresponding constant class instance.
- Gather the interpolation data
- Interpolate with `.interpolate()` or the overloaded call operator.

I guess the latter avoids us creating and destroying a class instance each time we interpolate, and we can remove the "determine the best scheme" process from the constructor. I just thought it'd be _safer_ to include this check in the constructor, as it means that the interpolation functions don't get cluttered with a bunch of `switch`es depending on which constant instance we need to point to?

Reply:
> The interp_scheme enum is there to make sure we can't request a scheme that doesn't exist, and just to make things a bit more readable than "use scheme 0".

Definitely agree that 'use scheme 0' is kinda useless! but I think having `best_interpolation_scheme` return a class instance ensures that it does exist though(?)

***

I like not creating a class for each call to `interpolate()`, also means we could get rid of some switch statements(?). Also it would be nice to remove the e..g `..(scheme - 6)...` expressions which aren't super readable.

***

Happy to approve/merge this is in and make a PR with a suggestion, if you think it's worth it

Reply:
> I like not creating a class for each call to interpolate(), also means we could get rid of some switch statements(?). Also it would be nice to remove the e..g ..(scheme - 6)... expressions which aren't super readable.


No worries, let me fix this here as merging it in is going will likely create problems down the line (the syntax for testing will likely change quite a bit).

So proposed changes are:
`interpScheme` class constructor just takes a value (enum, int, whatever) corresponding to the scheme that is to be used. We create constant instances of these classes (and give them suitable names), one for each scheme.
`best_interpolation_scheme(cells_in_direction, cell_id)` returns (a pointer to?) the constant class instance corresponding to the best interpolation scheme to use. This function is called whenever we want to interpolate.

I'll also address the `scheme-6` stuff by making the variable name a bit more informative and/or writing a method for `interpScheme` to extract this info!

This changes the previous approach where `best_interpolation_scheme(cells_in_direction, cell_id)` is built-into the `interpScheme` constructor, and an instance of the class is created & destroyed each time we interpolate.

Reply:
sounds good!

> No worries, let me fix this here as merging it in is going will likely create problems down the line 

cool cool

> interpScheme class constructor just takes a value (enum, int, whatever) corresponding to the scheme that is to be used. 

I would be inclined for the constructor to only take its associated 'value' (how good it is number). Then have a const instance for each scheme currently in `interp_scheme`. Also should have its index as an attribute (or some similar name) so we could get rid of the `..(scheme - 6)...` like things and have `scheme.index` instead (where `scheme` is a interpScheme instance). They could then each have associated `interpolate()` functions and not have to have lots of switch statements that currently exist in e.g. `bandlimited_interpolation(...)`

> (a pointer to?)

Definitely [return a reference](https://www.tutorialspoint.com/cplusplus/returning_values_by_reference.htm)

Reply:
> I would be inclined for the constructor to only take its associated 'value' (how good it is number).

Any strong reason for this preference? I personally prefer the other way - pass in the scheme identifier and then assign the value of the scheme, just because it's the scheme that provides the value, rather than the value dictating which scheme should be used?

Reply:
I thought it was the value which dictates what scheme will be used? to me that's the only thing that may change about an interpolation scheme. the index it acts on given some data and what it does are constant, the priority might change(?) e.g. if more interpolation schemes are added. I may be miss-understanding what you mean though 😕

Reply:
I'm using "value" rather ambiguously to be fair - maybe "preference" is a better word. From the interpolation document, here's how I understand how we should choose the interpolation scheme:
- If we can use BLi with 4 data points either side, use this. [This scheme has the highest "value"]
- If we can use BLi with 3 data points one side, and 5 the other, use this
- BLi with 2 one side / 6 other
- BLi with 1 one side / 7 other
- Cubic to centre of 4 points
- Cubic with 1 data point on one side / 3 on the other side [Lowest "value"]
The difficulty is that EG BLi with 3 data points "to the left" and 5 "to the right" has the same value as BLi with 5 data points "to the left" and 3 "to the right". So currently; "value" does not uniquely determine which "scheme" to be applied, but "scheme" does determine the "value".

Reply:
>  maybe "preference" is a better word.

definitely, or `priority`. I guess that number in addition to whether it is possible to use that scheme determines if it will be used

Reply:
Yeah, have rejigged the class now so hopefully things are a lot more readable (and compact). There's still an `enum scheme_value` to provide the `priority` of each scheme, but we now have constant class instances and `best_interpolation_scheme` returning references to them.
`interpolateTimeDomainE{x,y,z}` should be a lot more readable now too!

Reply:
love it. is `interpolateTimeDomainH{x,y,z}` getting the same treatment?

Reply:
Yep but will do this on another branch + PR. Sadly won't be _as_ nice to read due to the whole 2D-interpolation thing, but should still be much more readable